### PR TITLE
Fix a problem with accessing caller passed arguments on SystemV systems.

### PR DESCRIPTION
**Pull request by LLITCHEV**
The code was assuming that the register passed arguments will be always
spilled out to the outgoing arg area in the caller. Thus, if the RA
decided to home a register passed arg on the callee stack, accessing the
stack passed args will have a delta in the offset for the spilled arg.
This is incorrect. On SystemV systems the spilled register passed args are
homed on the callee stack. In case of frame register the extra delta for
the spilled arg is not needed. It is still needed for a non frame register
frame.
DDRs passed.
No Windows asm diffs.

### CODE REVIEW

**Comment by LLITCHEV** [comment]
@BruceForstall @CarolEidt @schellap 
Please, look at this.

File: src/jit/lclvars.cpp [hunk 1]
```diff
@@ -3710,7 +3709,7 @@ void Compiler::lvaFixVirtualFrameOffsets()
     
         if (doAssignStkOffs)
         {
-            varDsc->lvStkOffs += delta;
```
Comment (schellap) [inline src/jit/lclvars.cpp]: Unnecessary change.

File: src/jit/lclvars.cpp [hunk 2]
```diff
@@ -3861,7 +3863,11 @@ void Compiler::lvaAssignVirtualFrameOffsetsToArgs()
     {
         noway_assert(lclNum == info.compThisArg);
 #ifndef _TARGET_X86_
```
Comment (schellap) [inline src/jit/lclvars.cpp]: NIT: Would read better as:
#ifdef UNIX
Call()
#else
Call(foo)
#endif

in all places.

File: src/jit/compiler.cpp [hunk 3]
```diff
@@ -2055,7 +2055,9 @@ void                Compiler::compInitOptions(unsigned compileFlags)
             static ConfigMethodSet fJitDisasm;
             fJitDisasm.ensureInit(CLRConfig::INTERNAL_JitDisasm);
             if (fJitDisasm.contains(info.compMethodName, info.compClassName, info.compMethodInfo->args.pSig))
+            {
```
Comment (BruceForstall) [inline src/jit/compiler.cpp]: This matches our current coding conventions, but leaves everything else alone. It seems unrelated to your change. I wouldn't touch it unless you need to.

File: src/jit/compiler.h [hunk 4]
```diff
@@ -2274,7 +2274,11 @@ public :
 #endif // !LEGACY_BACKEND
 
     void                lvaAssignVirtualFrameOffsetsToArgs();
-    int                 lvaAssignVirtualFrameOffsetToArg(unsigned lclNum, unsigned argSize, int argOffs);
+    int                 lvaAssignVirtualFrameOffsetToArg(unsigned lclNum, unsigned argSize, int argOffs
+#ifdef UNIX_AMD64_ABI
```
Comment (BruceForstall) [inline src/jit/compiler.h]: You should look at AMD64_ARG and AMD64_FIRST_ARG in stdmacros.h. Maybe you could add a similar thing for UNIX_AMD64_ABI arguments.

File: src/jit/compiler.h [hunk 5]
```diff
@@ -2274,7 +2274,11 @@ public :
...
+                                                         , int& callerArgOffset
```
Comment (BruceForstall) [inline src/jit/compiler.h]: Is this argument in/out? We prefer int* to int&. Our conventions say that reference arguments should be 'const'.

File: src/jit/lclvars.cpp [hunk 6]
```diff
@@ -4044,12 +4086,14 @@ int Compiler::lvaAssignVirtualFrameOffsetToArg(unsigned lclNum, unsigned argSize
 #if defined(_TARGET_X86_)
         argOffs += sizeof(void *);
 #elif defined(_TARGET_AMD64_)
-#ifndef UNIX_AMD64_ABI
-        // The offset for args need not to be set for System V.
-        // No outgoing args for callees. The spilled args offsets will be set later.
-        varDsc->lvStkOffs = argOffs;
-        argOffs += sizeof(void *);
-#endif // !UNIX_AMD64_ABI
+#ifdef UNIX_AMD64_ABI
```
Comment (schellap) [inline src/jit/lclvars.cpp]: How would this work? Where are you taking care of the discontinuity in the frame for the return address and the ebp save.

File: src/jit/lclvars.cpp [hunk 7]
```diff
@@ -4044,12 +4086,14 @@ int Compiler::lvaAssignVirtualFrameOffsetToArg(unsigned lclNum, unsigned argSize
...
+        if (varDsc->lvOnFrame)
+#endif
+        {
+            // The offset for args need to be set only for the stack homed arguments for System V.
+            varDsc->lvStkOffs = argOffs;
+            argOffs += sizeof(void *);
+        }
 #elif defined(_TARGET_ARM64_)
         // Register arguments don't take stack space.
```
Comment (schellap) [inline src/jit/lclvars.cpp]: Also you might have to implement logic here for promoted struct as this is no longer true.

Comment (schellap) [inline src/jit/lclvars.cpp]: Also you might have to implement logic here for promoted struct as this is no longer true.

File: src/jit/emitxarch.cpp [hunk 8]
```diff
@@ -1721,7 +1721,10 @@ UNATIVE_OFFSET      emitter::emitInsSizeSV(size_t code, int var, int dsp)
 #endif
                 {
                     // Dev10 804810 - failing this assert can lead to bad codegen and runtime crashes
+#ifndef UNIX_AMD64_ABI
+                    // This can happen for SysteV calling convention when a registered passed argument is homed on the stack.
```
Comment (BruceForstall) [inline src/jit/emitxarch.cpp]: Can offs be positive? or should it just be <= 0 instead of < 0?

File: src/jit/lclvars.cpp [hunk 9]
```diff
@@ -4242,7 +4286,23 @@ int Compiler::lvaAssignVirtualFrameOffsetToArg(unsigned lclNum, unsigned argSize
         }
 #endif // _TARGET_ARM_
 
-        varDsc->lvStkOffs = argOffs;
+#ifdef UNIX_AMD64_ABI
```
Comment (schellap) [inline src/jit/lclvars.cpp]: It is not clear to me where the space for the spilled reg args is allocated and how you would skip the discontinuity in the arguments layout due to the ebp and the return address.

File: src/jit/emitxarch.cpp (hunk 8 above)
Comment (CarolEidt) [inline src/jit/emitxarch.cpp]: Typo: SystemV not SysteV

File: src/jit/compiler.h (hunk 4 above)
Comment (schellap) [inline src/jit/compiler.h]: This is a better approach than using:
#ifdef
#else
#endif

Similar to DEBUGARG

**Comment by schellap** [comment]
> Please use int* not int&, it makes the passing explicit at the caller.

**Comment by LLITCHEV** [comment]
@BruceForstall @CarolEidt @schellap 
 Please, take a look at this.
It contains changes addressing feedback of previous PR.
It also contains few more fixes:
1. Fix of a RA conservative assertion. In case where there are no
callee-save registers the RA was asserting this fact. It is a valid state
on System V machines for XMM registers. The RA does spill and frees a
register if the assertion doesn't trigger. The fix is to make the assert
allow non-callee-saved registers.
2. It fixes a problem with hardcoding registers when exception funclets
are generated. It uses platform independent RBM_ARG_0 macros now.
3. Fixes for calculating offsets for homeing arguments - the lvaAssignVirtualFrameOffsetToArg method is calculating a "raw" offset. The offset is fixed in lvaFixVirtualFrameOffsets "fixes" the "raw" offset by adjusting for the return address push, existence of frame pointer, frame padding, useage of alloca instructions, etc.
Tests and asm diffs runs are in progress.

**Comment by schellap** [comment]
LGTM

File: src/jit/lclvars.cpp [hunk 10]
```diff
@@ -4005,23 +4051,39 @@ void Compiler::lvaAssignVirtualFrameOffsetsToArgs()
...
 
-/*****************************************************************************
- *  lvaAssignVirtualFrameOffsetToArg() : Assign virtual stack offsets to an
- *  individual argument, and return the offset for the next argument.
- */
+//
+//  lvaAssignVirtualFrameOffsetToArg() : Assign virtual stack offsets to an
+//  individual argument, and return the offset for the next argument.
+//  Note: This method only calculates the initial offset of the stack passed/spilled arguments 
+//  (if any - the RA might decide to spill(home on the stack) register passed argumets, if rarely used.)
```
Comment (CarolEidt) [inline src/jit/lclvars.cpp]: typo - 'argumets' => 'arguments'

**Comment by CarolEidt** [comment]
LGTM, other than minor typo I mentioned
